agent-bootstrapper/src/com/thoughtworks/go/agent/bootstrapper/DefaultAgentLauncherCreatorImpl.java

```diff
@@ -16,80 +16,92 @@
+    private URLClassLoader urlClassLoader;
 
     public AgentLauncher createLauncher() {
-        inUseLauncher = createTempLauncherJar();
-        return (AgentLauncher) JarUtil.objectFromJar(inUseLauncher.getName(),
-                GO_AGENT_LAUNCHER_CLASS, GO_AGENT_LAUNCHER_LIB_DIR, AgentLauncher.class, AgentLaunchDescriptor.class);
-    }
-
-    private File createTempLauncherJar() {
-        File inUseLauncher;
+        createTempLauncherJar();
         try {
-            inUseLauncher = new File(UUID.randomUUID() + Downloader.AGENT_LAUNCHER);
-            FileUtils.copyFile(new File(Downloader.AGENT_LAUNCHER), inUseLauncher);
-        } catch (IOException e) {
+            return (AgentLauncher) loadClass(inUseLauncher, GO_AGENT_LAUNCHER_LIB_DIR, GO_AGENT_LAUNCHER_CLASS).newInstance();
+        } catch (Exception e) {
             throw new RuntimeException(e);
         }
-        LauncherTempFileHandler.startTempFileReaper();
-        return inUseLauncher;
     }
 
-    @Override
-    public void destroy() {
-        if (launcherDestroyed) {
-            return;
-        }
+    private Class<?> loadClass(File sourceJar, String libDirManifestKey, String classNameManifestKey) throws ClassNotFoundException {
+        String libDir = JarUtil.getManifestKey(sourceJar, libDirManifestKey);
+        String classNameToLoad = JarUtil.getManifestKey(sourceJar, classNameManifestKey);
 
-        recordCleanup();
+        Predicate<JarEntry> filter = jarEntry -> jarEntry.getName().startsWith(libDir + "/") && jarEntry.getName().endsWith(".jar");
 
-        attemptToCleanupMaxRetryTimes();
+        this.urlClassLoader = getClassLoaderFromJar(sourceJar, filter, getDepsDir(), DefaultAgentLauncherCreatorImpl.class.getClassLoader(), AgentLauncher.class, AgentLaunchDescriptor.class);
+        LOG.info("Attempting to load {} as specified by manifest key {}", classNameToLoad, classNameManifestKey);
+        return this.urlClassLoader.loadClass(classNameToLoad);
+    }
+
+    private URLClassLoader getClassLoaderFromJar(File aJarFile, Predicate<JarEntry> extractFilter, File outputTmpDir, ClassLoader parentClassLoader, Class... allowedClasses) {
+        List<File> urls = new ArrayList<>();
+        urls.add(aJarFile);
+        urls.addAll(JarUtil.extractFilesInLibDirAndReturnFiles(aJarFile, extractFilter, outputTmpDir));
+        ParentClassAccessFilteringClassloader filteringClassloader = new ParentClassAccessFilteringClassloader(parentClassLoader, allowedClasses);
+        return new URLClassLoader(toURLs(urls), filteringClassloader);
+    }
 
-        launcherDestroyed = true;
+    private void createTempLauncherJar() {
+        try {
+            inUseLauncher.getParentFile().mkdirs();
+            Files.copy(new File(Downloader.AGENT_LAUNCHER).toPath(), inUseLauncher.toPath());
+        } catch (IOException e) {
+            throw new RuntimeException(e);
+        }
+        LauncherTempFileHandler.startTempFileReaper();
```

Comment:
The current PR cleans up the temp-jars at two points, during a graceful shutdown and during bootstrapper restart. If the bootstrapper is an older version say 17.5 but the server was upgraded to 17.11, and a force agent shutdown happens, the temp-jars and the dependencies dir wouldn't be deleted. Since the boostrapper is an older version, these files would be left behind. 
`LauncherTempFileHandler` used to make an attempt at cleaning those up. If not temp-file-reaper, we need something else in place to clean up those files.